source/agora/test/GossipProtocol.d

```diff
@@ -0,0 +1,237 @@
+            node.sendMessage(msg);
+        }
+    }
+}
+
+/// Local Node for test
+class TestNode : TestAPI, IProtocolOwner {
+    private string name;
+    private TestNetwork network;
+    private GossipProtocol gossip;
+
+    public Hash lastReceivedMessage;
+    public int receivedMessageCount;
+
+    public this (string name) {
+        this.name = name;
+        this.network = new TestNetwork();
+        this.gossip = new GossipProtocol(this);
+
+        this.receivedMessageCount = 0;
+    }
+
+    /// PUT: /message
+    public bool setMessage(Hash msg) {
+		bool res = this.gossip.receiveMessage(msg);
+        if (res) {
+		    this.receivedMessageCount++;
+            this.lastReceivedMessage = msg;            
+        }
+        return res;
+    }
+
+    /// Set Romot Node
+    public void setRemotes (TestAPI [] api) {
+        this.network.setRemotes(api);
+    }
+
+    /// Send a message only to the node passed by the filter delegate.
+    public void sendMessage(Hash msg) {
+        this.network.sendMessage(msg);
+    }
+}
+
+unittest {
+    import std.stdio;
+
+    TestNode n1 = new TestNode("Node1");
+    TestNode n2 = new TestNode("Node2");
+    TestNode n3 = new TestNode("Node3");
+    TestNode n4 = new TestNode("Node4");
+    TestNode n5 = new TestNode("Node5");
+
+    n1.setRemotes([n2, n3]);
+    n2.setRemotes([n4, n5]);
+
+    Hash h = sha256Of("Message No. 1");
+    n1.setMessage(h);
+
+    assert(n1.receivedMessageCount == 1);
+    assert(n2.receivedMessageCount == 1);
+    assert(n3.receivedMessageCount == 1);
+    assert(n4.receivedMessageCount == 1);
+    assert(n5.receivedMessageCount == 1);
+
+    assert(n1.lastReceivedMessage == h);
+    assert(n2.lastReceivedMessage == h);
+    assert(n3.lastReceivedMessage == h);
+    assert(n4.lastReceivedMessage == h);
+    assert(n5.lastReceivedMessage == h);
+}
+
+unittest {
+    import std.stdio;
+
+    TestNode n1 = new TestNode("Node1");
+    TestNode n2 = new TestNode("Node2");
+    TestNode n3 = new TestNode("Node3");
+    TestNode n4 = new TestNode("Node4");
+    TestNode n5 = new TestNode("Node5");
+    TestNode n6 = new TestNode("Node6");
+    TestNode n7 = new TestNode("Node7");
+
+    n1.setRemotes([n2, n3]);
+    n2.setRemotes([n4, n5]);
+    n3.setRemotes([n6, n7]);
+
+    Hash h = sha256Of("Message No. 1");
+    n1.setMessage(h);
+
+    assert(n1.receivedMessageCount == 1);
+    assert(n2.receivedMessageCount == 1);
+    assert(n3.receivedMessageCount == 1);
+    assert(n4.receivedMessageCount == 1);
+    assert(n5.receivedMessageCount == 1);
+    assert(n6.receivedMessageCount == 1);
+    assert(n7.receivedMessageCount == 1);
+
+    assert(n1.lastReceivedMessage == h);
+    assert(n2.lastReceivedMessage == h);
+    assert(n3.lastReceivedMessage == h);
+    assert(n4.lastReceivedMessage == h);
+    assert(n5.lastReceivedMessage == h);
+    assert(n6.lastReceivedMessage == h);
+    assert(n7.lastReceivedMessage == h);
+}
+
+unittest {
+    import std.stdio;
```

Comment:
This import can also be removed.

Reply:
There are two more above which can be removed.

Reply:
Modified.

Reply:
@AndrejMitrovic 
Check Please~